### PR TITLE
Port segment store to flatbuffers

### PR DESCRIPTION
**Pull request by mavam**
This PR is a followup to #727 and ports the segment to store to use flatbuffers. Since the change involved actually changing the segment flatbuffer schema, the store is now completely stateless and initializes runtime state on startup.

### CODE REVIEW

File: libvast/src/segment_store.cpp [hunk 1]
```diff
@@ -352,12 +345,34 @@ void segment_store::inspect_status(caf::settings& dict) {
...
 
-segment_store::segment_store(path dir, uint64_t max_segment_size,
-                             size_t in_memory_segments)
-  : dir_{std::move(dir)},
-    max_segment_size_{max_segment_size},
-    cache_{in_memory_segments} {
-  // nop
+caf::error segment_store::register_segments() {
+  for (auto filename : directory{segment_path()})
+    if (auto err = register_segment(filename))
```
Comment (tobim) [inline libvast/src/segment_store.cpp]: We should probably _not_ break out of the loop here. An unrelated file that happens to be in this folder should not break the application.

File: libvast/src/segment_store.cpp [hunk 2]
```diff
@@ -352,12 +345,34 @@ void segment_store::inspect_status(caf::settings& dict) {
...
+      return err;
+  return caf::none;
+}
+
+caf::error segment_store::register_segment(const path& filename) {
+  auto chk = chunk::mmap(filename);
+  if (!chk)
+    return make_error(ec::filesystem_error, "failed to mmap chunk", filename);
+  auto s = fbs::GetSegment(chk->data());
```
Comment (tobim) [inline libvast/src/segment_store.cpp]: `s` needs to be checked for `nullptr`.

File: libvast/vast/fbs/segment.fbs [hunk 3]
```diff
@@ -17,14 +17,17 @@ table Segment {
   /// The version of the segment.
   version: Version;
 
+  /// The contained table slices.
```
Comment (tobim) [inline libvast/vast/fbs/segment.fbs]: Maybe we should add a magic word before the version so we can check that the file actually belongs to us.

File: libvast/src/segment_store.cpp [hunk 4]
```diff
@@ -352,12 +345,34 @@ void segment_store::inspect_status(caf::settings& dict) {
...
+  num_events_ += s->events();
+  auto segment_uuid = uuid{fbs::as_bytes<16>(*s->uuid())};
+  VAST_DEBUG_ANON(__func__, "found segment", segment_uuid);
```
Comment (tobim) [inline libvast/src/segment_store.cpp]: Why not
```suggestion
  VAST_DEBUG(this, "found segment", segment_uuid);
```

File: libvast/src/segment_store.cpp [hunk 5]
```diff
@@ -14,17 +14,17 @@
 #include "vast/segment_store.hpp"
 
 #include "vast/bitmap_algorithms.hpp"
+#include "vast/concept/printable/stream.hpp"
```
Comment (tobim) [inline libvast/src/segment_store.cpp]: Is this really necessary now?

File: libvast/vast/fbs/segment.fbs (hunk 3 above)
Comment (mavam) [reply libvast/vast/fbs/segment.fbs]: Good thought, it's actually already there. Have a look at the bottom of the file: the `file_identifier` directive does that natively.

File: libvast/src/segment_store.cpp (hunk 2 above)
Comment (mavam) [reply libvast/src/segment_store.cpp]: Why? The operation cannot fail. It just so happens that flatbuffers always returns a pointer.

File: libvast/src/segment_store.cpp (hunk 1 above)
Comment (mavam) [reply libvast/src/segment_store.cpp]: If we actually get an unrelated file, things break really bad when we call this a few lines before after mmaping the file:

```cpp
  auto s = fbs::GetSegment(chk->data());
```

I've added an extra check that bails out when the flatbuffer integrity is compromised.

File: libvast/src/segment_store.cpp (hunk 2 above)
Comment (dominiklohmann) [reply libvast/src/segment_store.cpp]: Maybe you can add `VAST_ASSERT(s)` in the line below for clarity.

Comment (mavam) [reply libvast/src/segment_store.cpp]: The thing is that all `Get*` can never be `nullptr`, by design. The function always returns a pointer value.

**Comment by mavam** [comment]
I've rebased the integration branch to master and then this topic branch onto the integration branch.

File: libvast/vast/fbs/segment.fbs (hunk 3 above)
Comment (tobim) [reply libvast/vast/fbs/segment.fbs]: I see.

File: libvast/src/segment_store.cpp (hunk 2 above)
Comment (tobim) [reply libvast/src/segment_store.cpp]: That was not clear to me, a pointer as return type signals the opposite to me as a reader after all.

File: libvast/src/segment_store.cpp [hunk 6]
```diff
@@ -357,6 +356,9 @@ caf::error segment_store::register_segment(const path& filename) {
   if (!chk)
     return make_error(ec::filesystem_error, "failed to mmap chunk", filename);
   auto s = fbs::GetSegment(chk->data());
+  auto verifier = fbs::make_verifier(chk);
+  if (!fbs::VerifySegmentBuffer(verifier))
+    return make_error(ec::format_error, "flatbuffer integrity check failed");
```
Comment (tobim) [inline libvast/src/segment_store.cpp]: Nit: the verification call to `fbs::GetSegment()` can be done after the verification.

Comment (tobim) [reply libvast/src/segment_store.cpp]: I'm also not sure this is a good idea. Consider the case of a large segment store (think of several TB): This verification means that we have to calculate the checksum for every segment during startup, which will probably take a long time.

Comment (mavam) [reply libvast/src/segment_store.cpp]: > Nit: the call to `fbs::GetSegment()` can be done after the verification.

Okay, I'll move it over.

> I'm also not sure this is a good idea. Consider the case of a large segment store (think of several TB): This verification means that we have to calculate the checksum for every segment during startup, which will probably take a long time.

It's not a checksum. It's just a verification that all offsets are within bounds and don't yield segfaults. It's not a full scan over the data. This is how the layout looks like:

[FBPerson01](https://github.com/mzaks/FlatBuffersSwift/wiki/FlatBuffers-Explained)

It's my understanding that verification only requires the initial table state.

Comment (tobim) [reply libvast/src/segment_store.cpp]: I see, this is good to go then.
